### ISSUE TITLE
`Optional.isEmpty()` not handled.

### ISSUE THREAD

**Issue by t-8ch** [typetools/checker-framework#4005]
It seems that `Optional.isEmpty()` is not handled by checker framework.
For that matter, no methods added to `java.util.Optional` in Java 9 or later are handled.

`.isEmpty()` is the exact inverse of the currently handled `.isPresent()`.
Switching the code to `!o.isPresent()` would be possible, but then another code checker complains that it should be rewritten as `o.isEmpty()`.

How are APIs that are not present in all JDK versions to be handled?
I could take a shot at implementing support for missing features for `j.u.Optional`.

## Commands

`javac -processor org.checkerframework.checker.optional.OptionalChecker OptionalDemonstration.java`

## Inputs

```java
import java.util.Optional;

public class OptionalDemonstration {
    public Object works() {
        Optional<String> o = makeOptional();
        
        if (o.isEmpty() || o.get().contains("foo")) {
            return 1;
        }
        return 2;
    }
    
    public Object doesNotWork() {
        Optional<String> o = makeOptional();

        if (!o.isPresent() || o.get().contains("foo")) {
            return 1;
        }
        return 2;
    }
    
    private static Optional<String> makeOptional() {
        return Optional.empty();
    }
}

```

## Output

```
OptionalDemonstration.java:9: error: [method.invocation.invalid] call to get() not allowed on the given receiver.
        if (o.isEmpty() || o.get().isEmpty()) {
                                ^
  found   : @MaybePresent Optional
  required: @Present Optional
1 error
```

## Expectations

No error should be reported. The same way `!Optional.isPresent()` works.

**Comment by wmdietl** [comment]
Thanks for the report!
Method `isEmpty()` needs a post-condition annotation `@EnsuresQualifierIf(result = false, expression = "this", qualifier = Present.class)` that tells the Optional Checker that if `isEmpty` returns false, the value is `@Present`.

Could you prepare a PR for this file:
https://github.com/typetools/jdk/blob/master/src/java.base/share/classes/java/util/Optional.java#L196

**Comment by t-8ch** [comment]
Will do.
Shouldn't there not also be support `OptionalVisitor`, for all features?

**Comment by mernst** [comment]
Thanks for noticing the missing annotations, and for the pull requests.  We appreciate it.

> For that matter, no methods added to `java.util.Optional` in Java 9 or later are handled.
> How are APIs that are not present in all JDK versions to be handled?

The `typetools/jdk` repository is a fork of JDK 11.  All methods that appear in JDK 11 should be annotated in the `typetools/jdk` repository.

For classes that appear in JDK 8 but not JDK 11, the [README](https://github.com/typetools/jdk/blob/master/README#L13) says:

> Annotations for classes that exist in JDK 8 but were removed in JDK 11 appear
> in jdk8.astub files in repository https://github.com/typetools/checker-framework/ .

> I could take a shot at implementing support for missing features for `j.u.Optional`.

That would be great.  Please do.  Thanks!

> Shouldn't there not also be support in OptionalVisitor, for all features?

What support is needed?  I think adding annotations to JDK methods is sufficient.

**Comment by t-8ch** [comment]
> > Shouldn't there not also be support in OptionalVisitor, for all features?
> 
> What support is needed? I think adding annotations to JDK methods is sufficient.

I assumed the logic in `OptionalVisitor` was a necessary counterpart for the annotations.
But actually it only generates warnings about suboptimal usage.
It currently does not support these new methods, so some warnings are missed.
But as these are not real errors and it seems not feasable to detect *all* forms of these suboptimal patterns it's fine I think.

**Comment by mernst** [comment]
Thanks for reminding me about the warnings.  I adjusted typetools/checker-framework#4015 to expand the warnings to new methods added in Java 11 (isEmpty and orElseThrow without an argument).
